secure-interpreter-webssh/handler.py

```diff
@@ -0,0 +1,651 @@
+        return args
+
+    def parse_encoding(self, data):
+        try:
+            encoding = to_str(data.strip(), 'ascii')
+        except UnicodeDecodeError:
+            return
+
+        if is_valid_encoding(encoding):
+            return encoding
+
+    def get_default_encoding(self, ssh):
+        commands = [
+            '$SHELL -ilc "locale charmap"',
+            '$SHELL -ic "locale charmap"'
+        ]
+
+        for command in commands:
+            try:
+                _, stdout, _ = ssh.exec_command(command,
+                                                get_pty=True,
+                                                timeout=1)
+            except paramiko.SSHException as exc:
+                logging.info(str(exc))
+            else:
+                try:
+                    data = stdout.read()
+                except socket.timeout:
+                    pass
+                else:
+                    logging.debug('{!r} => {!r}'.format(command, data))
+                    result = self.parse_encoding(data)
+                    if result:
+                        return result
+
+        logging.warning('Could not detect the default encoding.')
+        return 'utf-8'
+
+    def ssh_connect(self, args):
+        ssh = self.ssh_client
+        dst_addr = args[:2]
+        logging.info('Connecting to {}:{}'.format(*dst_addr))
+
+        try:
+            ssh.connect(*args, timeout=options.timeout)
+        except socket.error:
+            raise ValueError('Unable to connect to {}:{}'.format(*dst_addr))
+        except paramiko.BadAuthenticationType:
+            raise ValueError('Bad authentication type.')
+        except paramiko.AuthenticationException:
+            raise ValueError('Authentication failed.')
+        except paramiko.BadHostKeyException:
+            raise ValueError('Bad host key.')
+
+        term = self.get_argument('term', u'') or u'xterm'
+        chan = ssh.invoke_shell(term=term)
+        chan.setblocking(0)
+        worker = Worker(self.loop, ssh, chan, dst_addr)
+        worker.encoding = options.encoding if options.encoding else \
+            self.get_default_encoding(ssh)
+        return worker
+
+    def check_origin(self):
+        event_origin = self.get_argument('_origin', u'')
+        header_origin = self.request.headers.get('Origin')
+        origin = event_origin or header_origin
+
+        if origin:
+            if not super(IndexHandler, self).check_origin(origin):
+                raise tornado.web.HTTPError(
+                    403, 'Cross origin operation is not allowed.'
+                )
+
+            if not event_origin and self.origin_policy != 'same':
+                self.set_header('Access-Control-Allow-Origin', origin)
+
+    def head(self):
+        pass
+
+    """
+    認可判定を行い、アクセスを許可しないユーザーの場合400エラー
+    許可してるユーザーの場合、環境変数からhostname, usernameなどを取得しindex.htmlに渡す    
+    """
+    def get(self):
+        if not self.authZ():
+            raise tornado.web.HTTPError(400)
```

Comment:
認可が拒否された場合のエラーですので403ステータスで返却しましょうか
https://developer.mozilla.org/ja/docs/Web/HTTP/Status/403

Reply:
他の403エラーと区別できるようにメッセージも返しましょうか
```suggestion
            raise tornado.web.HTTPError(403, 'Forbidden')
```